Orbit.h

```diff
@@ -1,122 +1,104 @@
+#define ORBIT_H_
+
+#include <stdint.h>
+
+// Orbit API (header-only)
 //
-// DLL LOADING:
-// If you want to control the loading of Orbit64.dll, i.e. you don't want to
-// have to manually inject it into your application, use the OrbitAPI class.
-// Calling OrbitAPI::Connect(...) will load the dll and try to connect to the
-// Orbit instance specified in parameters.  Note that Orbit64.dll needs to be
-// next to your exectutable.  You can also change the code below
-// (OrbitAPI::Init()).
-//-----------------------------------------------------------------------------
-
-#define ORBIT_SCOPE(name) OrbitScope ORBIT_UNIQUE(ORB)(ORBIT_LITERAL(name))
-#define ORBIT_START(name) OrbitStart(ORBIT_LITERAL(name))
-#define ORBIT_STOP OrbitStop()
-#define ORBIT_SEND(ptr, num) OrbitSendData(ptr, num)
-
-//-----------------------------------------------------------------------------
-struct OrbitAPI {
-  static inline bool Connect(const char* a_Host, int a_Port = 1789);
-  static inline bool IsConnected();
-
- private:
-  static inline void Init();
-
-  typedef void (*InitRemote)(char*);
-  typedef bool (*GetBool)();
-
-  static inline HINSTANCE& GetHandle();
-  static inline InitRemote& GetInitRemote();
-  static inline GetBool& GetIsConnected();
-};
+// While the main feature of Orbit is its ability to dynamically instrument
+// functions, manual instrumentation is still possible using the macros below.
+// These macros call empty functions that Orbit dynamically instruments.
 
-//-----------------------------------------------------------------------------
-#define ORBIT_NOOP       \
-  static volatile int x; \
-  x;
-#define ORBIT_LITERAL(x) ("" x)
+// To disable manual instrumentation macros, set ORBIT_API_ENABLED to 0.
+#define ORBIT_API_ENABLED 1
 
-//-----------------------------------------------------------------------------
-#pragma optimize( \
-    "", off)  // On purpose to prevent compiler from stripping functions
-// NOTE: Do not use these directly, use above macros instead
-__declspec(noinline) inline void OrbitStart(const char*) { ORBIT_NOOP; }
-__declspec(noinline) inline void OrbitStop() { ORBIT_NOOP; }
-__declspec(noinline) inline void OrbitLog(const char*) { ORBIT_NOOP; }
-__declspec(noinline) inline void OrbitSendData(void*, int) { ORBIT_NOOP; }
-#pragma optimize("", on)
-
-//-----------------------------------------------------------------------------
-struct OrbitScope {
-  OrbitScope(const char* a_Name) { OrbitStart(a_Name); }
-  ~OrbitScope() { OrbitStop(); }
-};
+#if ORBIT_API_ENABLED
+
+// ORBIT_SCOPE: profile current scope.
+#define ORBIT_SCOPE(name) orbit::Scope ORBIT_UNIQUE(ORB)(ORBIT_LITERAL(name))
+
+// ORBIT_START/ORBIT_STOP: profile time span on a single thread.
+#define ORBIT_START(name) orbit::Start(ORBIT_LITERAL(name))
+#define ORBIT_STOP orbit::Stop()
+
+// ORBIT_START_ASYNC/ORBIT_STOP_ASYNC: profile time span across threads.
+#define ORBIT_START_ASYNC(name, id) orbit::StartAsync(ORBIT_LITERAL(name), id)
```

Comment:
If I understand this correctly, you need "id" to group the start and stop events together. Since start and stop potentially happen on different threads, will each "id" correspond to a unique track in the timeline view?

Reply:
For async time spans, we would still use the "name" as the track identifier.  "id" is there is to support having concurrent timespans on a single track.  A bit like what is being done for the GPU tracks.  Another example would be multiple async IO requests that are in flight at the same time. We need an "id" to match the start and end timestamps and we'd also want to visualize them in a single "IO" track.